stl/inc/ranges

```diff
@@ -624,6 +625,89 @@ namespace ranges {
         inline constexpr _Single_fn single;
     } // namespace views
 
+    // CLASS TEMPLATE ranges::istream_view
+    template <class _Ty, class _Elem, class _Traits>
+    concept _Stream_extractable = requires(basic_istream<_Elem, _Traits>& __is, _Ty& __t) {
+        __is >> __t;
+    };
+
+    // clang-format off
+    template <movable _Ty, class _Elem, class _Traits = char_traits<_Elem>>
+        requires default_initializable<_Ty> && _Stream_extractable<_Ty, _Elem, _Traits>
+    class basic_istream_view : public view_interface<basic_istream_view<_Ty, _Elem, _Traits>> {
+        // clang-format on
+    private:
+        class _Iterator {
+        private:
+            basic_istream_view* _Parent = nullptr;
+
+        public:
+            using iterator_concept = input_iterator_tag;
+            using difference_type  = ptrdiff_t;
+            using value_type       = _Ty;
+
+            _Iterator() = default;
+            constexpr explicit _Iterator(basic_istream_view& _Parent_) noexcept : _Parent{_STD addressof(_Parent_)} {}
+
+            _Iterator(const _Iterator&) = delete;
+            _Iterator(_Iterator&&)      = default;
+
+            _Iterator& operator=(const _Iterator&) = delete;
+            _Iterator& operator=(_Iterator&&) = default;
+
+            _Iterator& operator++() {
+#if _ITERATOR_DEBUG_LEVEL != 0
+                _STL_VERIFY(*this != default_sentinel, "cannot increment istream_view iterator at end of stream");
+#endif // _ITERATOR_DEBUG_LEVEL != 0
+                *_Parent->_Stream >> _Parent->_Val;
+                return *this;
+            }
+
+            void operator++(int) {
+                ++*this;
+            }
+
+            _Ty& operator*() const {
+#if _ITERATOR_DEBUG_LEVEL != 0
+                _STL_VERIFY(*this != default_sentinel, "cannot dereference istream_view iterator at end of stream");
+#endif // _ITERATOR_DEBUG_LEVEL != 0
+                return _Parent->_Val;
+            }
+
+            friend bool operator==(const _Iterator& _Left, default_sentinel_t) {
+                return _Left._Parent == nullptr || _Left._Parent->_Stream_at_end();
+            }
+        };
+
+        basic_istream<_Elem, _Traits>* _Stream = nullptr;
+        _Ty _Val                               = _Ty{};
+
+    public:
+        basic_istream_view() = default;
+        constexpr explicit basic_istream_view(basic_istream<_Elem, _Traits>& _Stream_) noexcept // strengthened
+            : _Stream{_STD addressof(_Stream_)} {}
+
+        constexpr auto begin() {
+            if (_Stream) {
+                *_Stream >> _Val;
+            }
+            return _Iterator{*this};
+        }
+
+        constexpr default_sentinel_t end() const noexcept {
+            return default_sentinel;
+        }
+
+        constexpr bool _Stream_at_end() const noexcept {

```

Comment:
Friendship limitations.

`_Stream `is a private member of `basic_istream_view`
`basic_istream_view::_Iterator` is a friend of `basic_istream_view` so the following code would be valid:
```cpp
bool operator==(const _Iterator& _Left, default_sentinel_t) const {
    return _Left._Parent == nullptr || !*_Left._Parent->_Stream;
}
```

However, the standard depicts this as 
```cpp
friend bool operator==(const _Iterator& _Left, default_sentinel_t) const {
    return _Left._Parent == nullptr || !*_Left._Parent->_Stream;
}
```

Now this function can access the private member `_Parent ` of `basic_istream_view::_Iterator` because it is a friend. However, that friendship does not apply transitively and a friend of `basic_istream_view::_Iterator` is not a friend of `basic_istream_view` and consequently has no access to `_Parent->_Stream`

I hope that makes sense.

Reply:
Great failure: I had intended to delete this comment before posting the review.

I reviewed this in our internal weekly code review class, and got to this before the `operator==`. I used it as an example of "It's not obvious what this public `_Ugly` function is doing here, let's leave a comment for now that we can come back and remove later if and when the purpose of this function becomes clear." (I've written enough range adaptors to know exactly why this function exists, but for teaching purposes it seemed best to play dumb. (It's a good idea to play dumb now and then so folks won't realize when you're not faking it.))

Of course we eventually got to `operator==` and all became clear, but I failed to come back and delete this comment.

Reply:
Wait you have code review classes. How awesome is that!

Can i know claim that my bugs are for instructional purposes?

Reply:
It's a new thing we're trying -- a code review by an experienced maintainer for less-experienced maintainers to learn from and we've opened it up to the larger team so they can get some insight into library/compiler features and considerations! :)

Reply:
I mean christmas decoration is going to pop up any time now, so could you record those sessions? 

I remember that we found that review is the main bottleneck and it would certainly help contributors to get more comfortable reviewing other PRs

Reply:
We really need to try to find a way to include contributors. I'll try to remember to bring it up in the weekly STL sync.